pyogrio/_io.pyx

```diff
@@ -1504,16 +1504,27 @@ def ogr_write(
     cdef OGRwkbGeometryType geometry_code
     cdef int err = 0
     cdef int i = 0
-    cdef int num_records = len(geometry)
-    cdef int num_fields = len(field_data) if field_data else 0
+    cdef int num_records = -1
+    cdef int num_field_data = len(field_data) if field_data is not None else 0
+    cdef int num_fields = len(fields) if fields is not None else 0
 
-    if len(field_data) != len(fields):
+    if num_fields != num_field_data:
         raise ValueError("field_data and fields must be same length")
 
-    if num_fields:
-        for i in range(1, len(field_data)):
-            if len(field_data[i]) != num_records:
-                raise ValueError("field_data arrays must be same length as geometry array")
+    if geometry is not None:
```

Comment:
It might be cleaner to validate that all fields have the same length first (if applicable) and then compare the length of the first field (if present) to the length of the geometry column (if present).  

```
if num_fields == 0 and geometry is None:
    raise ValueError("You must provide at least a geometry column or field data")

if num_fields > 0:
    num_records = len(field_data[0])
    for i in range (1, len(field_data)):
        if len(field_data[i) != num_records
            raise ValueError("all field_data arrays must be the same length")

if geometry is not None:
    if num_fields > 0:
        if len(geometry) != num_records:
            raise ValueError("field_data arrays must be same length as geometry array")
    else:
        num_records = len(geometry)
```

Reply:
This indeed increases readability!